fix(Java-SDK): handle generic service exceptions

texastony opened the pull request:
*Issue #, if available:*

SDK Exceptions that are not modeled by a service are treated as Opaque.
In Java, this lead to an IllegalStateException, instead of a NotAuthorized Exception.
This PR refactors the generated Java to capture un-modeled SDK exceptions,
such as NotAuthorized,
in the Opaque Error Types `obj` field,
and un-wrap them.

*Description of changes:*
- The Above
- Treat Markdown usage in documentation as warning instead of a danger; otherwise, current consumers (DB-ESDK) cannot pick up the latest Smithy-Dafny

Replaces https://github.com/smithy-lang/smithy-dafny/pull/464


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

texastony committed:
fix(Java-SDK): handle generic service exceptions

texastony committed:
fix(Java-SDK): handle generic service exceptions

texastony committed:
test(Java-SDKv2): handle generic service in opaque exception

texastony committed:
fix: unsupport documnatation trait as warning, not danger

robin-aws commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToDafnyAwsV2.java:
```diff
@@ -727,11 +693,24 @@ MethodSpec generateConvertOpaqueError() {
         subject.nativeNameResolver.baseErrorForService(),
```
AFAICT the only difference between `generateConvertOpaqueServiceError` and `generateConvertOpaqueError` is the parameter type (the service's bass exception type vs. `Exception`). Couldn't we just have the latter in that case?

texastony replied:
While they behave identically today, they, ultimately, should not.
I will address via a comment.

Developers MAY treat un-modled KMS Exceptions differently than truly unexpected exceptions.
i.e: NotAuthorized vs Java Heap out of Memory 

That both of these exceptions end up as Opaque, today, is a miss.

It does not allow Dafny developers to treat the two separately.

robin-aws commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToNativeAwsV2.java:
```diff
@@ -511,4 +512,66 @@ protected MethodSpec modeledUnion(final UnionShape shape) {
+        subject.nativeNameResolver.baseErrorForService()
+      )
+      .addStatement(
+        "return ($T) $L.$L",
```
This doesn't end up behaving any differently from the next block AFAICT, since it only casts to a more specific exception type, but that distinction is lost since since the return type of the method is just `RuntimeException`.

texastony replied:
We agreed to keep it.
I have documented why.

robin-aws commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/nameresolver/AwsSdkNativeV2.java:
```diff
@@ -84,6 +85,13 @@ public AwsSdkNativeV2(final ServiceShape serviceShape, final Model model) {
         Map.entry("com.amazonaws.dynamodb", "DynamoDbException"),
         Map.entry("com.amazonaws.s3", "S3Exception")
       );
+    // These short names are used for convince in docs and errors messages.
```
How about using [`smithy.api#title`](https://smithy.io/2.0/spec/documentation-traits.html#smithy-api-title-trait) instead? That's defined for all relevant services, likely all AWS services in fact. And it should be the most human-readable and marketing-friendly name for a service, such as "Amazon Simple Storage Service" or "Amazon DynamoDB"

texastony replied:
Done.

robin-aws commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToNativeAwsV2.java:
```diff
@@ -511,4 +512,66 @@ protected MethodSpec modeledUnion(final UnionShape shape) {
+        Dafny.datatypeConstructorIs("Some")
+      )
+      .addStatement(
+        "final $T message = $L($L.$L.$L)",
```
Perhaps this should stringify the opaque error value as well, like the last case below.

texastony replied:
Done.

robin-aws reviewed (changes requested):
Definitely a big improvement to make sure a native error value is unwrapped so it round trips through Dafny, and it seems doable to ensure the same thing in any language (assuming we can safely hold onto such a value with a Dafny `object` reference, but that's an orthogonal issue).

It seems to be we can simplify a bit to avoid more cases that only work well for known services though - see other comments.

texastony committed:
fix: Danger for /// documentation only

As compared to Danger also for Markdown,
which breaks current Smithy-Dafny consumers.

texastony committed:
chore(Java-SDK): address PR comments

texastony committed:
chore(Java-SDK): spelling and grammar

texastony committed:
chore(Java): prettier

texastony committed:
chore(Java-SDK): Unit Testing Changes

seebees reviewed (approved):
Looks good.
I like adding the comments to the generated code. Though I do wish that there was a way to manage the comments and have a single string for them. However, I'm not sure how to do this cleanly and not overly complicate the code generation code...

robin-aws reviewed (approved):
🚀
